tests/functional/test_model.py

```diff
@@ -0,0 +1,133 @@
+"""Contains tests for the DatabaseModel backends"""
+
+import logging
+import os
+from typing import List, Optional
+
+from databases import Database
+from pydantic import Field
+import pytest
+from sqlalchemy import desc
+
+from pynocular.backends.context import backend
+from pynocular.backends.memory import MemoryDatabaseModelBackend
+from pynocular.backends.sql import SQLDatabaseModelBackend
+from pynocular.model import DatabaseModel
+
+
+@pytest.fixture()
+async def postgres_backend():
+    """Fixture that yields a Postgres backend
+
+    Yields:
+        postgres backend
+
+    """
+    db_host = os.environ.get("DB_HOST", "localhost")
+    db_user_name = os.environ.get("DB_USER_NAME", os.environ.get("USER", "postgres"))
+    db_user_password = os.environ.get("DB_USER_PASSWORD", "")
+    test_db_name = os.environ.get("TEST_DB_NAME", "test_db")
+
+    maintenance_connection_string = f"postgres://{db_user_name}:{db_user_password}@{db_host}:5432/postgres?sslmode=disable"
+    db_connection_string = f"postgresql://{db_user_name}:{db_user_password}@{db_host}:5432/{test_db_name}?sslmode=disable"
+
+    async with Database(maintenance_connection_string) as db:
+        try:
+            await db.execute(f"DROP DATABASE IF EXISTS {test_db_name}")
+            await db.execute(f"CREATE DATABASE {test_db_name}")
+        except Exception as e:
+            logging.info(str(e))
+
+    async with Database(db_connection_string) as db:
+        await db.execute(
+            "CREATE TABLE IF NOT EXISTS things (id SERIAL PRIMARY KEY, name TEXT)"
+        )
+        try:
+            yield SQLDatabaseModelBackend(db)
+        finally:
+            await db.execute("DROP TABLE things")
+
+    async with Database(maintenance_connection_string) as db:
+        try:
+            await db.execute(f"DROP DATABASE IF EXISTS {test_db_name}")
+        except Exception as e:
+            logging.info(str(e))
+
+
+@pytest.fixture()
+async def memory_backend():
+    """Fixture that yields an in-memory backend
+
+    Returns:
+        in-memory backend
+
+    """
+    return MemoryDatabaseModelBackend()
+
+
+class Thing(DatabaseModel, table_name="things"):
+    """A test database model"""
+
+    id: Optional[int] = Field(primary_key=True)
+    name: str = Field()
+
+
+async def _run_tests():
+    """Run tests agnostic to the backend"""
+    things: List[Thing] = await Thing.select()
```

Comment:
since you didn't define a backend what is this `select()` using?

Reply:
This function is a test helper that runs in context of a backend (see test_postgres and test_memory) below. The `Thing` class is a `DatabaseModel` so it's calling `DatabaseModel.select()` which returns a list of database model instances.

Reply:
I missed that this was a helper function